File: rsa.c
// Standard libraries
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

// OpenSSL libraries
#include <openssl/bn.h>
#include <openssl/err.h>
#include <openssl/evp.h>
#include <openssl/pem.h>
#include <openssl/rand.h>
#include <openssl/sha.h>

// Constants
#define RSA_KEY_SIZE 2048 // bits
#define SEED_SIZE 32 // bytes

// Struct for RSA key pair
typedef struct rsa_key_pair
{
    BIGNUM *n;
    BIGNUM *e;
    BIGNUM *d;
} RSA_KEY_PAIR;

/**
 * @file rsa.c
 *
 * @brief This file contains the implementation of the RSA algorithm
 *
 * @date 2023-11-16
 *
 * @author Ana Raquel Neves Vidal (118408)
 * @author Simão Augusto Ferreira Andrade (118345)
 *
 */

/**
 * @brief Function to check if A is a subarray of B
 *
 * @param A reference array A
 * @param a_size size of A
 * @param B reference array B
 * @param b_size size of B
 *
 * @return 1 if A is a subarray of B, 0 otherwise
 */
int pattern_found(uint8_t A[], int a_size, uint8_t B[], int b_size)
{
    if (a_size > b_size)
    {
        fprintf(stderr,"A is longer than B\n");
        return 0;
    }

    for (int i = 0; i <= b_size - a_size; i++)
    {
        int match = 1;
        for (int j = 0; j < a_size; j++)
        {
            if (B[i + j] != A[j])
            {
                match = 0;
                break; // mismatch
            }
        }

        if (match)
        {
            return 1;
        }
    }

    return 0; // A is not a subarray of B
}

/**
 * @brief Function to use the generator to produce a pseudo-random stream of bytes
 *
 * @param stream a pointer to an array where the bytes will be allocated
 * @param stream_size the size of the array
 * @param seed value that will start the generator
 */
void generate_pseudo_random_stream(uint8_t *stream, int stream_size, uint8_t seed[SEED_SIZE])
{
    uint8_t hash_output[SHA256_DIGEST_LENGTH];

    for (int index = 0; index < stream_size; index += SHA256_DIGEST_LENGTH)
    {
        SHA256(seed, SEED_SIZE, hash_output); // hash the seed

        // Copy the hash output to the stream
        int bytes_to_copy = (index + SHA256_DIGEST_LENGTH <= stream_size) ? SHA256_DIGEST_LENGTH : stream_size - index;
        memcpy(stream + index, hash_output, bytes_to_copy);

        memcpy(seed, hash_output, SHA256_DIGEST_LENGTH); // seed = hash_output
    }
}

/**
 * @brief This function generates arbirtary a pseudo-random byte stream
 *
 * @param size The size of the byte stream to be generated
 * @param password The password to be used
 * @param confusion_string The confusion string to be used
 * @param iterations The number of iterations to be used
 * @param bytes The output byte array
 *
 */
void randgen(int size, const char *password, const char *confusion_string, int iterations, uint8_t *bytes)
{
    uint8_t key_derivator[SEED_SIZE + strlen(confusion_string)];
    uint8_t seed[SEED_SIZE];
    uint8_t confusion_pattern[strlen(confusion_string)];

    // generate the key derivator
    if (PKCS5_PBKDF2_HMAC(password, strlen(password), (const unsigned char *)confusion_string, strlen(confusion_string), iterations, EVP_sha256(), SEED_SIZE + strlen(confusion_string), key_derivator) != 1)
    {
        fprintf(stderr, "Error generating key derivator\n");
        exit(1);
    }

    // get the seed and the confusion pattern from the key derivator
    memcpy(seed, key_derivator, SEED_SIZE);
    memcpy(confusion_pattern, key_derivator + SEED_SIZE, strlen(confusion_string));

    for (int iteration = 0; iteration < iterations; iteration++)
    {
        uint8_t temp_buffer[size];
        while (1)
        {
            generate_pseudo_random_stream(temp_buffer, size, seed);

            if (pattern_found(confusion_pattern, strlen(confusion_string), temp_buffer, size))
            {
                break;
            }
        }

        memcpy(bytes, temp_buffer, size);
        memcpy(seed, bytes + size - SEED_SIZE, SEED_SIZE);
    }
}

/**
 * @brief This function encodes a byte array in base64
 * 
 * @param input  The byte array to be encoded
 * @param length The length of the byte array
 * @return char* The encoded byte array
 */
char *base64_encode(const unsigned char *input, int length)
{
    BIO *bmem, *base64;
    BUF_MEM *buffer_pointer;

    base64 = BIO_new(BIO_f_base64());
    bmem = BIO_new(BIO_s_mem());
    base64 = BIO_push(base64, bmem);
    BIO_write(base64, input, length);
    BIO_flush(base64);
    BIO_get_mem_ptr(base64, &buffer_pointer);

    char *buffer = (char *)malloc(buffer_pointer->length + 1);
    memcpy(buffer, buffer_pointer->data, buffer_pointer->length);
    buffer[buffer_pointer->length] = 0;

    BIO_free_all(base64);

    return buffer;
}

/**
 * @brief This function stores the RSA key pair in a PEM file.
 *
 * @param key_pair The RSA key pair to be stored
 * @param private_key_filename The name of the file to store the key pair
 * @param public_key_filename The name of the file to store the key pair
 */
void storekey(RSA_KEY_PAIR key_pair, const char *private_key_filename, const char *public_key_filename)
{
    FILE *private_key_file = fopen(private_key_filename, "wb");
    FILE *public_key_file = fopen(public_key_filename, "wb");

    // Private key
    if (!private_key_file)
    {
        fprintf(stderr, "Error opening file\n");
        exit(1);
    }

    fprintf(private_key_file, "-----BEGIN PRIVATE KEY-----\n");

    int private_key_data_len = BN_num_bytes(key_pair.n) + BN_num_bytes(key_pair.d);

    unsigned char *private_key_data = (unsigned char *)malloc(private_key_data_len);
    if (!private_key_data)
    {
        fprintf(stderr, "Error allocating memory\n");
        exit(1);
    }

    BN_bn2bin(key_pair.n, private_key_data);
    BN_bn2bin(key_pair.d, private_key_data + BN_num_bytes(key_pair.n));

    char *private_key_data_base64 = base64_encode(private_key_data, private_key_data_len);
    fprintf(private_key_file, "%s", private_key_data_base64);

    free(private_key_data);

    fprintf(private_key_file, "-----END PRIVATE KEY-----\n");

    fclose(private_key_file);

    if (!public_key_file)
    {
        fprintf(stderr, "Error opening file\n");
        exit(1);
    }

    // Public key 
    fprintf(public_key_file, "-----BEGIN PUBLIC KEY-----\n");

    int public_key_data_len = BN_num_bytes(key_pair.n) + BN_num_bytes(key_pair.e);

    unsigned char *public_key_data = (unsigned char *)malloc(public_key_data_len);
    if (!public_key_data)
    {
        fprintf(stderr, "Error allocating memory\n");
        exit(1);
    }

    BN_bn2bin(key_pair.n, public_key_data);
    BN_bn2bin(key_pair.e, public_key_data + BN_num_bytes(key_pair.n));

    char *public_key_data_base64 = base64_encode(public_key_data, public_key_data_len);
    fprintf(public_key_file, "%s", public_key_data_base64);

    free(public_key_data);

    fprintf(public_key_file, "-----END PUBLIC KEY-----\n");

    fclose(public_key_file);
}

/**
 * @brief This function generates an RSA key pair from a Pseudo-Random Byte Stream
 *
 * @param bytes The pseudo-random byte array that will be used to generate the key pair
 *
 * @return The RSA key pair
 */
RSA_KEY_PAIR rsagen(uint8_t *bytes)
{
    RSA_KEY_PAIR key_pair = {NULL, NULL, NULL};

    BIGNUM *p = BN_new();
    BIGNUM *q = BN_new();
    BIGNUM *n = BN_new();
    BIGNUM *e = BN_new();
    BIGNUM *d = BN_new();
    BN_CTX *ctx = BN_CTX_new();
    if (!p || !q || !n || !e || !d || !ctx)
    {
        fprintf(stderr, "Error initializing BIGNUM variables\n");
        goto cleanup;
    }

    // Divide the pseudo-random bytes in two halves
    uint8_t *bytes_p = bytes;
    uint8_t *bytes_q = bytes + RSA_KEY_SIZE / 16;


    if (!BN_bin2bn(bytes_p, RSA_KEY_SIZE / 16, p)) // p value
    {
        fprintf(stderr, "Error setting P value\n");
        goto cleanup;
    }

    if (!BN_bin2bn(bytes_q, RSA_KEY_SIZE / 16, q)) // q value
    {
        fprintf(stderr, "Error setting Q value\n");
        goto cleanup;
    }

    while (!BN_is_prime_ex(p, BN_prime_checks, ctx, NULL) || !BN_is_prime_ex(q, BN_prime_checks, ctx, NULL) || BN_cmp(p, q) == 0)
    {
        if (!BN_add_word(p, 1))
        {
            fprintf(stderr, "Error incrementing P\n");
            goto cleanup;
        }

        if (!BN_add_word(q, 1))
        {
            fprintf(stderr, "Error incrementing Q\n");
            goto cleanup;
        }
    }

    if (!BN_mul(n, p, q, ctx)) // n = p * q
    {
        fprintf(stderr, "Error calculating modulus\n");
        goto cleanup;
    }

    if (!BN_set_word(e, RSA_F4)) // 2^16 + 1
    {
        fprintf(stderr, "Error setting exponent\n");
        goto cleanup;
    }

    if (!BN_mod_inverse(d, e, n, ctx)) // d = e^-1 mod n
    {
        fprintf(stderr, "Error calculating private exponent\n");
        goto cleanup;
    }

    key_pair.n = n;
    key_pair.e = e;
    key_pair.d = d;

// Free variables
cleanup:
    BN_free(p);
    BN_free(q);

    if (ctx)
        BN_CTX_free(ctx);

    if (ferror(stderr))
        exit(EXIT_FAILURE);

    return key_pair;
}

int main(int argc, char **argv)
{
    if (argc != 4) // Usage => ./rsa <password> <confusion_string> <iterations>
    {
        fprintf(stderr, "Usage: ./rsa <password> <confusion_string> <iterations>\n");
        exit(1);
    }

    const char *password = argv[1];
    const char *confusion_string = argv[2];
    int iterations = atoi(argv[3]);

    // Generate random bytes
    uint8_t bytes[RSA_KEY_SIZE / 8];
    randgen(RSA_KEY_SIZE / 8, password, confusion_string, iterations, bytes);

    // Generate RSA key pair
    RSA_KEY_PAIR key_pair = rsagen(bytes);

    storekey(key_pair, "private_key.pem", "public_key.pem");

    return 0;
}

File: rsa.java
import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.Arrays;
import java.util.Base64;
import javax.crypto.spec.PBEKeySpec;
import javax.crypto.SecretKeyFactory;
import java.io.*;

/**
 * This file contains the implementation of the D-RSA algorithm in Java.
 *
 * @author Ana Raquel Neves Vidal (118408)
 * @author Simão Augusto Ferreira Andrade (118345)
 *
 */
public class RSA {
    public static final String PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA256";
    public static final int SIZE_STREAM = 2048; // Tamanho da chave RSA
    public static final int ITERATIONS = 10; // Número máximo de iterações
    public static final String PASSWORD = "password";
    public static final String CONFUSION_STRING = "abc";
    public static final int SEED =  32; // Tamanho da seed em bytes
    public static final int SHA256_DIGEST_LENGTH = 32; // Tamanho do output do SHA-256 em bytes
    public static final BigInteger e = BigInteger.valueOf(65537); // Valor de e
    
    public static void main(String[] args) throws Exception {
        
        // Generate random stream of bytes
        byte[] random_stream_bytes = randgen(PASSWORD, CONFUSION_STRING, ITERATIONS, SIZE_STREAM / 8);

        // Generate the key pair
        BigInteger[][] keyPair = generateKeyPair(random_stream_bytes);

        // Store the key pair in files
        storeKeyPair(keyPair);
    }

    /**
     * Generates a random stream of bytes
     * 
     * @param streamSize The size of the stream
     * @param seed The seed
     * 
     * @return The random stream of bytes
    */
    public static byte[] generateRandomBytes(int streamSize, byte[] seed) {
        try {
            MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
            byte[] stream = new byte[streamSize];

            for (int index = 0; index < streamSize; index += SHA256_DIGEST_LENGTH) {
                // Use SHA-256 to hash the seed
                byte[] hashOutput = sha256.digest(seed);

                // Copy the hash output to the stream
                int bytesToCopy = (index + SHA256_DIGEST_LENGTH > streamSize) ? streamSize - index : SHA256_DIGEST_LENGTH;
                
                // Copy the hash output to the stream
                System.arraycopy(hashOutput, 0, stream, index, bytesToCopy);

                // Update the seed with the hash output for the next iteration
                System.arraycopy(hashOutput, 0, seed, 0, SHA256_DIGEST_LENGTH);

            }
            
            return stream;
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace(); // Handle the exception appropriately
            return null;
        }
    }

    public static String encodeToBase64(byte[] data) {
        return Base64.getEncoder().encodeToString(data);
    }

    public static boolean verificarPadrao(byte[] A, byte[] B) {

        if(A.length > B.length){
            System.out.println("A is bigger than B");
            return false;
        }

        for (int i = 0; i <= B.length - A.length; i++) {
            int j;
            for (j = 0; j < A.length; j++) {
                if (B[i + j] != A[j]) {
                    break;
                }
            }
            if(j == A.length){
                return true;
            }
        }   
        return false;
    }
    
    public static byte[] randgen(String password, String confusionString, int iterations, int size)
            throws NoSuchAlgorithmException, InvalidKeySpecException {

        // Derivar a chave usando PBKDF2 com a senha, o salt que é numa maneira inicial, a confusion string e o número de iterações
        PBEKeySpec spec = new PBEKeySpec(password.toCharArray(), confusionString.getBytes(), iterations, (SEED + confusionString.length()) * 8); //256 bits para a chave
        SecretKeyFactory skf = SecretKeyFactory.getInstance(PBKDF2_ALGORITHM);
        byte[] key_derivator = skf.generateSecret(spec).getEncoded();

        byte[] bootstrapSeed = new byte[SEED];
        bootstrapSeed = Arrays.copyOfRange(key_derivator, 0, SEED);

        byte[] confusionPattern = new byte[confusionString.length()];
        confusionPattern = Arrays.copyOfRange(key_derivator, SEED, SEED + confusionString.length());

        byte[] stream = new byte[size];
        for(int iteration = 0; iteration < iterations; iteration++){
           while(true){
                stream = generateRandomBytes(size, bootstrapSeed);

                if(verificarPadrao(confusionPattern, stream)){
                    break;
                }
           }
           
           bootstrapSeed = Arrays.copyOfRange(stream, stream.length - SEED, stream.length); // pass the stream to the bootstrap seed
        }
        return stream;
    }

    /**
     * Generates a key pair
     * 
     * @param random_stream_bytes The random stream of bytes
     * 
     * @return The key pair
    */
    public static BigInteger[][] generateKeyPair(byte[] random_stream_bytes) throws Exception {
        // Dividir o stream aleatório em duas partes p e q
        byte[] p_bytes = Arrays.copyOfRange(random_stream_bytes, 0, random_stream_bytes.length / 2);
        byte[] q_bytes = Arrays.copyOfRange(random_stream_bytes, random_stream_bytes.length / 2, random_stream_bytes.length);

        // tornar o p e q como bignumbers e certificar que são primos
        BigInteger p = new BigInteger(1, p_bytes).nextProbablePrime();
        BigInteger q = new BigInteger(1, q_bytes).nextProbablePrime();

        // Calcular o n
        BigInteger n = p.multiply(q);

        // Calcular o phi
        BigInteger phi = p.subtract(BigInteger.ONE).multiply(q.subtract(BigInteger.ONE));

        // Calcular o d
        BigInteger d = e.modInverse(phi);
        

        // criar a chave publica pub=(n,e)
        BigInteger[] public_key = {n, e};

        // criar a minha chave privada priv=(n,d)
        BigInteger[] private_key = {n, d};

        //colocar num array o pub e priv
        BigInteger[][] keyPair = {public_key, private_key};

        return keyPair;
    }

    /**
     * Stores the public key in a file
     * 
     * @param keyPair The key pair to store
    */
    public static void storeKeyPair(BigInteger[][] keyPair) throws IOException {
        // Get the public key
        BigInteger[] public_key = keyPair[0];

        // passar o n e o e para um array de bytes
        byte[] n_bytes = public_key[0].toByteArray();
        byte[] e_bytes = public_key[1].toByteArray();

        // juntar o n e o e num array de bytes
        byte[] public_key_bytes = new byte[n_bytes.length + e_bytes.length];
        System.arraycopy(n_bytes, 0, public_key_bytes, 0, n_bytes.length);
        System.arraycopy(e_bytes, 0, public_key_bytes, n_bytes.length, e_bytes.length);

        // codificar o array de bytes para base64
        String public_key_base64 = encodeToBase64(public_key_bytes);

        // imprimir a chave publica
        System.out.println("------BEGIN PUBLIC KEY------\n" + public_key_base64 + "\n------END PUBLIC KEY------\n");

        // Get the private key
        BigInteger[] private_key = keyPair[1];

        // passar o n e o d para um array de bytes
        byte[] n_bytes_priv = private_key[0].toByteArray();
        byte[] d_bytes_priv = private_key[1].toByteArray();

        // juntar o n e o d num array de bytes
        byte[] private_key_bytes = new byte[n_bytes_priv.length + d_bytes_priv.length];
        System.arraycopy(n_bytes_priv, 0, private_key_bytes, 0, n_bytes_priv.length);
        System.arraycopy(d_bytes_priv, 0, private_key_bytes, n_bytes_priv.length, d_bytes_priv.length);

        // codificar o array de bytes para base64
        String private_key_base64 = encodeToBase64(private_key_bytes);

        // imprimir a chave privada
        System.out.println("------BEGIN PRIVATE KEY------\n" + private_key_base64 + "\n------END PRIVATE KEY------\n");

        // Store the public key in a file
        try (FileOutputStream outputStream = new FileOutputStream("public_key.pem")) {
            outputStream.write(public_key_base64.getBytes());
        }catch(Exception e){
            System.out.println("Erro ao escrever no ficheiro");
        }

        // Store the private key in a file
        try (FileOutputStream outputStream = new FileOutputStream("private_key.pem")) {
            outputStream.write(private_key_base64.getBytes());
        }catch(Exception e){
            System.out.println("Erro ao escrever no ficheiro");
        }
    }
}
